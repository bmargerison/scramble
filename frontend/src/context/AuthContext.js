import createDataContext from './createDataContext';
import axios from "axios";

const authReducer = (state, action) => {
  switch (action.type) {
    case 'signout':
      return {token: null, userId: ''};
    case 'signin':
    case 'signup':
      return {
        token: action.payload.token,
        userId: action.payload.userId,
      };
    default:
      return state;
  }
};

const signup = dispatch => {
  return ({username, email, password}) => {
    axios
      .post("http://localhost:3000/users", {
        username: username,
        email: email,
        password: password
      })
      .then((res) => {
        console.log(res.data);
      })
      .catch((err) => window.alert(err.response.data.message));
    axios
      .post("http://localhost:3000/users/login", {
        email: email,
        password: password
      })
      .then(async (res) => {
        dispatch({
          type: 'signin',
          payload: {
            token: res.data.token,
            userId: res.data.user._id
          },
        })
      })
      .catch((err) => window.alert(err.response.data.message));
  };
};

const signin = dispatch => {
  return ({email, password}) => {
    axios
      .post("http://localhost:3000/users/login", {
        email: email,
        password: password
      })
      .then(async (res) => {
        dispatch({
          type: 'signin',
          payload: {
            token: res.data.token,
            userId: res.data.user._id
          },
        })
      })
      .catch((err) => window.alert(err.response.data.message));
  };
};

const signout = dispatch => {
  return () => {
    dispatch({type: 'signout'});
  };
};

export const {Provider, Context} = createDataContext(
  authReducer,
  {signin, signout, signup},
  {token: null, userId: ''},
);