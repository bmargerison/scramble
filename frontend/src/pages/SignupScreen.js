import React, { useState, useContext } from "react";
import { View, Button, StyleSheet, Text, TextInput, TouchableOpacity} from 'react-native';
import {Context as AuthContext} from '../context/AuthContext';
import {AppStyles} from '../AppStyles';

const SignupScreen = ({ navigation }) => {
  const [username, setUsername] = useState();
  const [email, setEmail] = useState();
  const [password, setPassword] = useState();
  const {state, signup} = useContext(AuthContext);

  return (
    <View style={styles.container}>
      <Text style={[styles.title, styles.leftTitle]}>Create new account</Text>
      <View style={styles.InputContainer}>
        <TextInput
          style={styles.body}
          placeholder="Username"
          onChange={(e) => setUsername(e.target.value)}
          value={username}
          placeholderTextColor={AppStyles.color.grey}
          underlineColorAndroid="transparent"
        />
      </View>
      <View style={styles.InputContainer}>
        <TextInput
          style={styles.body}
          placeholder="E-mail Address"
          onChange={(e) => setEmail(e.target.value)}
          value={email}
          placeholderTextColor={AppStyles.color.grey}
          underlineColorAndroid="transparent"
        />
      </View>
      <View style={styles.InputContainer}>
        <TextInput
          style={styles.body}
          placeholder="Password"
          secureTextEntry={true}
          onChange={(e) => setPassword(e.target.value)}
          value={password}
          placeholderTextColor={AppStyles.color.grey}
          underlineColorAndroid="transparent"
        />
      </View>
      <TouchableOpacity
        style={styles.loginContainer}
        onPress={() => signup({username, email, password})}>
        <Text style={styles.loginText}>Sign up</Text>
      </TouchableOpacity>
      <Text style={styles.loginLink}
        onPress={() => navigation.navigate('Signin')}>
        Already have an account? Login here
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
  },
  title: {
    fontSize: AppStyles.fontSize.title,
    fontWeight: 'bold',
    color: AppStyles.color.tint,
    marginTop: 50,
    marginBottom: 50,
  },
  leftTitle: {
    alignSelf: 'stretch',
    textAlign: 'left',
    marginLeft: 20,
  },
  loginContainer: {
    width: AppStyles.buttonWidth.main,
    backgroundColor: AppStyles.color.tint,
    borderRadius: AppStyles.borderRadius.main,
    padding: 10,
  },
  loginText: {
    color: AppStyles.color.white,
    alignSelf: "center",
  },
  InputContainer: {
    width: AppStyles.textInputWidth.main,
    marginBottom: 30,
    borderWidth: 1,
    borderStyle: 'solid',
    borderColor: AppStyles.color.grey,
    borderRadius: AppStyles.borderRadius.main,
  },
  body: {
    height: 42,
    paddingLeft: 20,
    paddingRight: 20,
    color: AppStyles.color.text,
  },
  loginLink: {
    marginTop: 20,
    marginBottom: 20,
    color: AppStyles.color.blue,
  },
});

export default SignupScreen;
