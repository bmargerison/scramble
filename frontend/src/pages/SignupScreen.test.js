/**
 * @jest-environment jsdom
 */

import React from 'react';
import SignupScreen from './SignupScreen';
import {Provider as AuthProvider} from '../context/AuthContext.js';
import { shallow, mount } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

describe('signup screen', () => {

  it('should match to snapshot', () => {
    const component = shallow(    
      <AuthProvider>
        <SignupScreen/>
      </AuthProvider>
    )
    expect(component).toMatchSnapshot()
  });

  it('should have a username, email and password field', () => {
    const component = mount(    
      <AuthProvider>
        <SignupScreen />
      </AuthProvider>
    )
    expect(component.find('TextInput').first().prop('placeholder')).toEqual('Username')
    expect(component.find('TextInput').at(2).prop('placeholder')).toEqual('Email')
    expect(component.find('TextInput').at(4).prop('placeholder')).toEqual('Password')
  });

  it('should have a signup button', () => {
    const component = mount(    
      <AuthProvider>
        <SignupScreen/>
      </AuthProvider>
    )
    expect(component.find('Button').first().prop('title')).toEqual('Sign Up')
  });

});