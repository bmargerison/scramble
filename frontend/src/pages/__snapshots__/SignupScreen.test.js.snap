// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onChange={[Function]}
    placeholder="Username"
    placeholderTextColor="white"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "#AFEEEE",
        "borderRadius": 14,
        "color": "white",
        "fontSize": 18,
        "fontWeight": "500",
        "height": 55,
        "margin": 10,
        "padding": 8,
        "width": 350,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onChange={[Function]}
    placeholder="Email"
    placeholderTextColor="white"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "#AFEEEE",
        "borderRadius": 14,
        "color": "white",
        "fontSize": 18,
        "fontWeight": "500",
        "height": 55,
        "margin": 10,
        "padding": 8,
        "width": 350,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onChange={[Function]}
    placeholder="Password"
    placeholderTextColor="white"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Object {
        "backgroundColor": "#AFEEEE",
        "borderRadius": 14,
        "color": "white",
        "fontSize": 18,
        "fontWeight": "500",
        "height": 55,
        "margin": 10,
        "padding": 8,
        "width": 350,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Sign Up
      </Text>
    </View>
  </View>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "blue",
      }
    }
  >
    Already have an account? Login here
  </Text>
</View>
`;
